File: SortingAlgorithms/SelectionSort.ts
/**
 * Selection Sort implementation
 * Time Complexity : Quadratic O(n^2)
 * Space Complexity : Constant O(1)
 */

let unsorted_list = [64, 34, 25, 12, 22, 11, 90, 5];

function selectionSort(arr: number[]): number[] {
  let n = arr.length;

  if (n <= 1) return arr;

  for (let i = 0; i < n; i++) {
    let min_index = i;

    for (let j = i + 1; j < n; j++) {
      if (arr[min_index] > arr[j]) min_index = j;
    }
    let tmp = arr[i];
    arr[i] = arr[min_index];
    arr[min_index] = tmp;
  }

  return arr;
}
console.log(selectionSort(unsorted_list));

File: Algorithms/SortingAlgorithms/SelectionSort.py

'''
Selection sort implementation using nested for loop
Big-P
    Time Complexity  : quadratic O(n^2)
    Space Complexity : constant O(1)
'''

unsorted_list = [64,34,25,12,22,11,90,5]

def selection_sort(ls):
    n = len(ls)

    if n <= 1:
        return ls
    
    for i in range(n):
        min_index = i

        for j in range(i + 1, n):
            if ls[j] < ls[min_index]:
                min_index = j

        ls[i], ls[min_index] = ls[min_index], ls[i]

    return ls

print(selection_sort(unsorted_list))